### ISSUE TITLE
Wexflow Record use.

### ISSUE THREAD

**Issue by janborup** [aelassas/Wexflow#234]
Hi 

I'm trying to find out how this Webflow Records works.

I have created a record - Check !
I have uploadet af document for the record - Check !

Now I want to Approve this record - How do I do this - the "Approved" button is deactivated. ?
Now I want to assign this record to a user - How do I do this - The "Assign to" is deactivated ?

![image](https://user-images.githubusercontent.com/6866443/81597328-db946f80-93c5-11ea-9894-e3a9a0241527.png)

**Comment by aelassas** [comment]
@janborup You must create a new approval workflow in the designer and drag and drop ApproveRecord task in the designer, select your record in the task settings and select the user assigned to the record in the task settings. Finally, go to Approvals tab and kick off your approval workflow in order to assign the record to the user.

**Comment by janborup** [comment]
Okay..
- I created a Workflow for this a test record with an "ApproveRecord".
- I assigned it to a user.
- I Logged in as this user, and there was a notification
![image](https://user-images.githubusercontent.com/6866443/81606609-adb62780-93d3-11ea-8ef2-abb54213b6f2.png)

Now there is a lot of notifications, but all with the same line.:
`"An approval process on the record Test has started. You must update that record by adding new file versions. You can also add comments on that record."`

But now what to do next ?
What to approve, and where to do it ?

![image](https://user-images.githubusercontent.com/6866443/81607157-84e26200-93d4-11ea-9088-c2e7b255192c.png)
The Approve-tab is empty:

![image](https://user-images.githubusercontent.com/6866443/81607707-66c93180-93d5-11ea-925e-06f0ca2ff479.png)
The Record-tab has the task, but the approve is diabled.

![image](https://user-images.githubusercontent.com/6866443/81607893-b14aae00-93d5-11ea-8a94-96656b13f8fc.png)
Same when openning the Record, the "Approved" is diabled.

Where do the user approve this ?

/Jan

**Comment by aelassas** [comment]
@janborup The user can only add files and comments on the record. It's the manager who approves the record. Generally, the user who started the approval workflow.

**Comment by janborup** [comment]
Hi Again

I don't get what i'm doing wrong.
If I in a Workflow assign a record to a user like "RecordCreator", then this user cannot approve the record only add documents, is that right ?
The record is allready uploadet with a document, how do i sent this to a approver, and how to the approver see the attached documents ?

/Jan

**Comment by aelassas** [comment]
@janborup If the user assigned to the record adds a document, the approver will recieve a notification. The approver can see the attachments by opening the record in Records tab.

Here is a brief overview of an approval flow:

1. Create a record (The manager can delete records. The user assigned to a record can modify the record by uploading new file versions and adding comments but he cannot delete the record he has not created. The user assigned to a record can only delete the records he created.).
2. Create a workflow that takes as input the record and the user assigned to that record (A super-Admin or Admin user in Wexflow).
3. When the manager (A Super-Admin user in Wexflow) kicks off an approval workflow, a notification is sent to the user assigned to the record saying that he was assigned to work on that record.
4. Once the user assigned to the record modifies the record (by uploading a new file version of the record for example), a notification is sent to the manager saying that he have to check that record.
5. Once the manager checks the record he can add comments and if so, a new notification is sent to the user assigned to the record.
6. If the manager deletes the record, the user assigned to the record is notified.
7. If the record's due date is reached, the user assigned to the record and the manager are notified.
8. If the manager approves the record, a notification is sent to the user assigned to the record saying that the record was approved.
9. If the manager rejects the record, a notification is sent to the user assigned to the record saying that the record was rejected.
10. The approval can be stopped by the manager.

A non-manager user can also create records and approval workflows but he cannot access the records he did not create and the approval workflows he did not create. He can only edit the records created by him or assigned to him. He cannot delete the records he did not create but he can delete the records he created. If a non-manager user modifies a record assigned to him, the manager of the record receives a notification and an email. If the manager modifies a record assigned to a user, the user assigned to the record receives a notification and an email.

**Comment by janborup** [comment]
Okay.. Now I got the Record-WorkFlow (**RWF**) to work, but I think there is serveral flaws, if this  should be functional and more user friendly.

First the issues i experienced:
- If you are a SuperAdministrator, you can create a RWF, with an Administrator assigned as approver, but the SuperAdministrator can Approve the same record self, and there i no logging about who has approved the record.

There is a lot of steps that should be done, just to do a simple RWF approvement.
- If you are a approver of a record, the "Approvals"-tab needs to be refactored :
--  You cannot see what record or document you are approving
-- Not see records details
-- Not see who an when has been notified for this record 
-- Logging of what happens with the record.
-- You can see all this in all the other tabs, but it you just have 10 records to approve, it will take a long time to do just the approval.

- If you are a record creator, all the Record / RWF creation will take long time to do.
- For next version I world recommend a new task for RecordWorkFlowImport with a template to import, like the way doing the email-template.
- I have attached an example how this could med the RWF-creation muck simpler.
-- For each document in a ImportFolder, this task should create a new RWF with the <RecordWorkflow> af WF template
 
[RecordWorkflowImports.xml.txt](https://github.com/aelassas/Wexflow/files/4617605/RecordWorkflowImports.xml.txt)

I hope this could help simplify this RWF 
Say if I should do something on this

/Cheers

**Comment by aelassas** [comment]
@janborup I understand what you want to do. You want to create a hot folder for records and auto create and auto start the approval workflow automatically once the record's file is imported in the database from the hot folder. I will work on that in the v5.7.

**Comment by janborup** [comment]
@aelassas 
Yep.... The Hotfolder RWF AutoCreate function will save a lot of manual time on creation.

My concerns was also about the Approval-tab, because if you are a manager approver, you actually only want it to be very simple and informative in the "Approval-Tab/page": 
1. Click to see the next document and who has assigned it (Only started jobs)
2. Maybe set at comment - Shortcut: Save
3. Click "Approve" and Next record/document in one task.
This way you can approve 5-6 documents in a minute, and all other Wexflow functions is normally inrelevant for a normal approver. 

Suggestions:
- New Userprofile : Approver
- Only see started jobs, and assigned for the approver.
- Max One click or shortcut to open the record document
- Button [Approve] - [Reject] - [Approve+Next] shortcuts.
- Button [Record] to see/change the record information.
- Button [Log] to see logged information.
- Button [Notifications] to see notification information for the record.
This could be a time-saver for a normal approver. 

Future request:
- A record finished flag on the record.
- Multible Approvers
- Assign extra Approver in startet workflow.
- Auto delete RWF at finished, but keep log information in records.
- A suggestion could be a "SmartApproval" page, optimized for mobile

Hope some of this comes in you pipeline considerations.

/Jan

**Comment by aelassas** [comment]
@janborup I added multiple approvers feature and a hot folder for importing records which by default is C:\Wexflow\Records\\_HotFolder. These features will be available in the v5.7.

If you want to test importing records, pull the latest version and add this setting in Wexflow.xml:

```xml
<Setting name="recordsHotFolder" value="C:\Wexflow\Records\_HotFolder" />
```

**Comment by janborup** [comment]
Hi @aelassas 

I have testet the _hotFolder to -> Record, and this works fine. 👍 - but no RWF is autocreated for this.

I think you missed the [RecordWorkflowImports.xml.txt](https://github.com/aelassas/Wexflow/issues/234#issuecomment-627509938) template file I suggested above, because this takes care of multible Hotfolders, with multible Workflow templates, because normally VendorInvoices, Serviceinvoices, DepartmentInvoices, Timesheets, Travelbills and more, because normally is not the same department/manager approving all documents, and if you only have one _Hotfolder in Wexflow.xml, it would be difficult to manage this goal.
Another solution the TemplateFileImport could be a WF with:
1. FileSystemWatcher task (For a record hotfolder)
2. RecordDocumentImporter task (Created records for files in hotfolder) 
3. CreateWorkFlow task (Settings - Record: FromImporter - AssignedTo: Admin, Wexflow - LaunchType: StartUp)  

I cannot get the multible Approvers to work, but maybe I doing it wrong.
- I have created a WF with 2 ApproveRecord-tasks, and 2 Approval-tasks (Only the first approver work)
- I have created 2 WF with 1 ApproveRecord-task, and 1 Approval-task, one for Admin and one for Wexflow (Only the first approver admin is showed up in records).

/Jan

**Comment by aelassas** [comment]
@janborup I like the TemplateFileImport:

1. FileSystemWatcher task (Watching record's hotfolder)
2. ApprovalRecordCreator task (Creates records from files in hotfolder)
3. ApprovalWorkflowCreator task (Settings - Record: From Importer - AssignedTo:  a user from settings - LaunchType: StartUp)

I will try to work on RecordCreator and ApprovalWorkflowCreator tasks in the v5.7.

For multiple approvers, here is what you should do to make it work:

1. Create an approval workflow with only one ApproveRecord task on a record R.
2. Start the first workflow using the user admin.
3. Create a second approval workflow with only one ApproveRecord task on the same record record R.
4. Create a super admin user named approver for example.
5. Login as approver user and start the second workflow.
7. Approve the second workflow using the user approver.
8. Login as admin and approve the first workflow.

You will see a table in records tab once you click on a record containing all approvers.

I also added a label for approved records, just pull the latest version and you will see that on approved records in records tab.

**Comment by janborup** [comment]
Hi

Regarding the multible Approvers, I have done all above Step 1 - 7 by the book, and nothing fails.

Still in the Record, it's only the first approver for "Admin" visible in the list, not the second one. 
![image](https://user-images.githubusercontent.com/6866443/82125012-461d2500-97a3-11ea-8e2f-83bf0e9bdf00.png)

**Comment by aelassas** [comment]
Did you do the following steps:

3. Create a second approval workflow with only one ApproveRecord task on
the same record record R.
4. Create a super admin user named *approver*.
5. Login as *approver* user and start the second workflow.
6. *Approve the second workflow using the user approver.* (At this point,
go to records tab and click on your record R to see if *approver* user was
added)
7. Login as admin and approve the first workflow. (At this point, go to
records tab and click on your record R to see if *admin* user was added)

The two approval workflows must be different and approved by different
users.

I have just tested it two times and it works fine on my end.



Le sam. 16 mai 2020 à 16:37, Jan Borup Coyle <notifications@github.com> a
écrit :

**Comment by janborup** [comment]
Okay

I only got it to work one time, but only if approver from the Second WF approves first, and the approver from first WF approves last, and that is even if both WF is started from the beginning.

The problem is also if both approvers is SuperAdmins, then in Approval-tab, both jobs is visible for both users for approval, and still in Approval-tab / job.... you have absolutely no idea about what you are approving, even the Record name, Record description or Document is visible or trackable in Approval-tab. 
If you as manager have several Records to approve, you have no chance to quess what WF it belongs to.

**Comment by aelassas** [comment]
@janborup If the two approval workflows are running in parallel, you can approve the first workflow first or second it is the same thing. To check that the approvers were added, once you start the second workflow go to records tab, refresh the page and click on the record, you will see the two approvers with approved flag set to false (unchecked). Once, you start approving the record, you can refresh the page and click on the record to see the approvers getting updated.

Regarding the Approvals tab issue, maybe naming the approval workflow with the user's username and record's name may solve this. By doing this, the logged in manager can easily find his created approval workflows and approve/reject them. But for multiple managers, the approvers must be super admin Wexflow users. For the moment, there's no way to do it otherwise. 

An admin user can be an approver and can only see his approval workflows, this solves the issue but the admin user as an approver can only approve/reject his created records. He cannot share records with other admin users. Multiple admin approvers is not possible for the moment because admin users cannot share the same record.

**Comment by janborup** [comment]
Hi @aelassas 

Just testet ApprovalRecordCreator-task, and is works fine 👌

/Cheers

**Comment by janborup** [comment]
A lille test problem with the ApprovalRecordCreator

I create this workflow:
![image](https://user-images.githubusercontent.com/6866443/82153462-dcb31a00-9867-11ea-9850-5d322d3f9d3e.png)

Then i drop 2 or more files in the import folder.
Only the some of files is Record-created.

Maybe it's a delay problem. !!!

**Comment by aelassas** [comment]
@janborup All fixed. Can you pull the latest version and try again? Also, you don't need FilesRemover task because files are moved from FileSystemWatcher folder to Records folder in ApprovalRecordsCreator task.

**Comment by janborup** [comment]
Hi @aelassas 

I just fetch latest, YES - now it works like a charm. 😎

Then it only the last part of the automatisation.. the 3. ApprovalWorkflowCreator task
 
I will test this when you are ready.

/Cheers

**Comment by aelassas** [comment]
@janborup I have just added ApprovalWorkflowsCreator task. Just pull the latest version, update your TasksNames.json and TasksSettings.json, add ApprovalWorkflowsCreator into your workfow, set approver and assignedTo settings, update FileSystemWatcher task and test. 

ApprovalWorkflowsCreator creates an approval workflow for the record and starts it. This task must always be after ApprovalRecordsCreator task because it takes as input the records created by ApprovalRecordsCreator task.

**Comment by janborup** [comment]
**Hi @aelassas** 

**Very good Job done, i have testet this RecordFileImportFlow both in Windows.Net and in .Net Core, and in both environments the fileflow works great. 😎👍**

I had a little stange issue when testing, because if an Approver [Reject] a record, the WF stops and nothing happens.
If the WF is then again started, but then [Stopped] by the Appover, it creates a second Approver for the creator (Admin) of the WF, not the assignedBy ??????
![image](https://user-images.githubusercontent.com/6866443/82256402-91c40000-9956-11ea-8685-22df112ce4a6.png)
If the WF is then again started, and [Approved] by the approver, the created approver (Admin)m then disappers again, and only the Approver is registered as Approver.
![image](https://user-images.githubusercontent.com/6866443/82257060-ac4aa900-9957-11ea-820c-1c57330c130a.png)
Does it make sense - maybe this is supposed to !!!

**Comment by aelassas** [comment]
> If the WF is then again started, and [Approved] by the approver, the created approver (Admin)m then disappers again, and only the Approver is registered as Approver.

@janborup Yes, this is done by design. If you stop the approval workflow and start it again all the approved approvers are deleted and the new ones are added during the approval flow.

Did you test deleteWorkflowOnApproval setting in FileSystemImport? You don't have to configure anything, deleteWorkflowOnApproval is by default set true in ApprovalWorkflowsCreator task. Just pull the latest version, update your TasksSettings.json and test.

**Comment by janborup** [comment]
Yes. Sure..

Ii tested the deleteWorkflowOnApproval In both versions, and it works like a charm. 🤩

I think we got the goal for this issue, and then we take a more user friendly Approval overview, in another topic.

Thanks - good work. 😁
